packages/clarity-js/src/layout/extract.ts

```diff
@@ -0,0 +1,67 @@
+    for (let v of value) {
+        let valid = true;
+        let filter = v.length === 4 ? v[2] : null;
+        switch (filter) {
+            case RegionFilter.Url: valid = !!top.location.href.match(new RegExp(v[3])); break;
+            case RegionFilter.Javascript: valid = !!evaluate(v[3]); break;
+        }
+        if (valid) { root.querySelectorAll(v[1]).forEach(e => region.observe(e, v[0].toString())); }
+    }
+}
+
+export function metrics(root: ParentNode, value: Metric[]): void {
+    for (let v of value) {
+        let source = v.length >= 3 ? v[1] : null;
+        let scale = v.length === 4 ? v[3] : 1;
+        switch (source) {
+            case Extract.Text:  root.querySelectorAll(v[2]).forEach(e => { metric.max(v[0], num((e as HTMLElement).innerText, scale)); }); break;
+            case Extract.Attribute: root.querySelectorAll(`[${v[2]}]`).forEach(e => { metric.max(v[0], num(e.getAttribute(v[2]), scale, false)); }); break;
+            case Extract.Javascript: metric.max(v[0], evaluate(v[2], Constant.Number) as number); break;
+        }
+    }
+}
+
+// The function below takes in a variable name in following format: "a.b.c" and safely evaluates its value in javascript context
+// For instance, for a.b.c, it will first check window["a"]. If it exists, it will recursively look at: window["a"]["b"] and finally,
+// return the value for window["a"]["b"]["c"].
+function evaluate(variable: string, type: string = null, base: Object = window): any {
+    let parts = variable.split(Constant.Dot);
+    let first = parts.shift();
+    if (base && base[first]) {
+        if (parts.length > 0) { return evaluate(parts.join(Constant.Dot), type, base[first]); } 
+        let output = type === null || type === typeof base[first] ? base[first] : null;
+        return output;
+    }
+    return null;
+}
+
+function num(text: string, scale: number, localize: boolean = true): number {
+    try {
+        // Reference: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat
+        let lang = document.documentElement.lang;
+        if (Intl && Intl.NumberFormat && lang && localize) {
+            text = text.replace(digitsWithCommaRegex, Constant.Empty);
+            // Infer current group and decimal separator from current locale
+            let group = Intl.NumberFormat(lang).format(11111).replace(formatRegex, Constant.Empty);
+            let decimal = Intl.NumberFormat(lang).format(1.1).replace(formatRegex, Constant.Empty);
+            
+            // Prase number using inferred group and decimal separators
+            return Math.round(parseFloat(text
+                .replace(new RegExp('\\' + group, 'g'), Constant.Empty)
+                .replace(new RegExp('\\' + decimal), Constant.Dot)
+            ) * scale);
+        }
+        // Fallback to en locale
+        return Math.round(parseFloat(text.replace(digitsRegex, Constant.Empty)) * scale);
+    } catch { return 0; }
```

Comment:
CAn we not put a number if we fail here. Cleaner to check where this is called and return undefined if we can't parse it.

Reply:
Yes right makes complete sense. No idea why I had it at 0 but fixed it to return null.